src/database/mod.rs

```diff
@@ -158,6 +168,73 @@ pub trait Database: BatchOperations {
     ///
     /// It should insert and return `0` if not present in the database
     fn increment_last_index(&mut self, keychain: KeychainKind) -> Result<u32, Error>;
+
+    /// Delete UTXOs provided as params, if `spent_utxos` is empty it deletes all spent
+    /// UTXOs in the database.
+    fn del_spent_utxos(
+        &mut self,
+        spent_outpoints: Option<Vec<OutPoint>>,
+    ) -> Result<Vec<LocalUtxo>, Error> {
+        if let Some(spent_outpoints) = spent_outpoints {
+            let deleted_utxos = spent_outpoints
+                .iter()
+                .filter_map(|out| self.del_utxo(out).transpose())
+                .collect::<Result<Vec<_>, _>>()?;
+            Ok(deleted_utxos)
+        } else {
+            let deleted_utxos = self
+                .iter_utxos()?
+                .iter()
+                .filter(|utxo| utxo.is_spent)
+                .filter_map(|out| self.del_utxo(&out.outpoint).transpose())
+                .collect::<Result<Vec<_>, _>>()?;
+            Ok(deleted_utxos)
+        }
+    }
+
+    /// Delete UTXOs based on the number of confirmations or
+    /// threshold size (number of spent utxos in DB) defined in [`DelCriteria`]
+    fn del_spent_utxos_by_criteria(
+        &mut self,
+        criteria: DelCriteria,
+    ) -> Result<Vec<LocalUtxo>, Error> {
+        let spent_utxos: Vec<LocalUtxo> = self
+            .iter_utxos()?
+            .into_iter()
+            .filter(|utxo| utxo.is_spent)
+            .collect();
+        if let Some(threshold_size) = criteria.threshold_size {
+            // if the threshold size is smaller than current size, delete everything
+            if threshold_size < spent_utxos.len() as u64 {
+                //delete all spent utxos
+                return self.del_spent_utxos(None);
+            }
+        }
+
+        let mut filtered_outpoints: Vec<OutPoint> = vec![];
+
+        // get utxos that matches criteria and delete them
+        match self.get_sync_time()? {
+            Some(sync_time) => {
+                for spent_utxo in spent_utxos {
+                    let tx_details = self.get_tx(&spent_utxo.outpoint.txid, false)?;
+                    let height = tx_details
+                        .and_then(|details| details.confirmation_time)
+                        .map(|conf| conf.height);
+                    let confirmations = criteria.confirmations;
+                    if height.is_some()
+                        && confirmations.is_some()
+                        && ((sync_time.block_time.height - height.unwrap())
+                            > confirmations.unwrap())
+                    {
+                        filtered_outpoints.push(spent_utxo.outpoint);
+                    }
+                }
+                Ok(self.del_spent_utxos(Some(filtered_outpoints))?)
+            }
+            None => Ok(vec![]),
+        }
+    }
```

Comment:
I am not much sure on the criteria part much.. If the wallet dev wants to delete utxos upon some criteria they should be able to do that on their own in their code. And such criteria in practice can be very diverse. It doesn't help for us to anticipate them and provide defaults. More options at such low level database API is more confusing..

But if it has to stay, I would expect a behavior from the structure something like below..

Both the criteria are defined as options. So they are both free to be set independent of each other. So when both is set I would expect an `and` of them, and when one is `None` it should not have any effect.

So there code here should do something like this, instead of deleting    per each criteria.
 - Choose all utxos to delete by confirmation criteria.
 - Apply threshold criteria on remaining utxos and get how many more to delete, if any. If some, select them by decreasing confirmation time from utxo list.
 - Compile both of the two list together. 
  - Delete those utxos..